bin/verdicts.py

```diff
@@ -0,0 +1,452 @@
+    - first_unknown[testgroup]: first child whose verdict is not (yet) known; None if none exists
+
+    - _verdict[testnode]: the verdict at the given testnode, or None. In particular,
+        verdict['.'] is the root verdict, sometimes called final verdict or submission verdict.
+        Should not be directly set; use __setitem__ on the Verdict object instead.
+    - _unknowns[testgroup]: iterator over the children that do not (yet) have a verdict,
+        in lexicographic order
+    """
+
+    def unknowns_iterator(self, node):
+        """Yield the node's (yet) unknown children in lexicographic order."""
+        for child in sorted(self.children[node]):
+            if self._verdict[child] is not None:
+                continue
+            yield child
+
+    def __init__(self, testcase_list: list[str]):
+        testcases = set(testcase_list)
+        testgroups: set[str] = set(str(path) for tc in testcases for path in Path(tc).parents)
+        self._verdict: dict[str, Verdict | None] = {g: None for g in testcases | testgroups}
+        self.duration: dict[str, int | None] = {g: None for g in testcases}
+
+        self.children: dict[str, set[str]] = {node: set() for node in testgroups}
+        for node in testcases | testgroups:
+            if node != '.':
+                parent = str(Path(node).parent)
+                self.children[parent].add(node)
+        self.first_error: dict[str, str | None] = {node: None for node in testgroups}
+        self.num_unknowns: dict[str, int] = {node: len(self.children[node]) for node in testgroups}
+        self._unknowns = {node: self.unknowns_iterator(node) for node in testgroups}
+        self.first_unknown: dict[str, str | None] = {
+            node: next(self._unknowns[node]) for node in testgroups
+        }
+
+    def is_testgroup(self, node) -> bool:
+        """Is the given testnode name a testgroup (rather than a testcase)?
+        This assumes nonempty testgroups.
+        """
+        return node in self.children
+
+    def is_testcase(self, node) -> bool:
+        """Is the given testnode name a testcase (rather than a testgroup)?
+        This assumes nonempty testgroups.
+        """
+        return node not in self.children
+
+    def __setitem__(self, testcase, verdict: str | Verdict, duration=None):
+        """Set the verdict of the given testcase (implying possibly others)
+
+        verdict can be given as a Verdict or as a string using either long or
+        short form ('ACCEPTED', 'AC', or Verdict.ACCEPTED).
+        """
+
+        if isinstance(verdict, str):
+            verdict = from_string(verdict)
+        if verdict == Verdict.ACCEPTED and duration is not None:
+            self.duration[testcase] = duration
+        self._set_verdict_for_node(testcase, verdict)
+
+    def __getitem__(self, testnode) -> Verdict | None:
+        return self._verdict[testnode]
+
+    def salient_testcase(self) -> str:
```

Comment:
Related to my [previous comment](https://github.com/RagnarGrootKoerkamp/BAPCtools/pull/358#discussion_r1527228362): we probably want the salient test case to be the slowest test case when any sub-verdict is TLE as well? Note that TLE may not be the first rejected verdict, especially with non-lazy judging. So, for the verdicts `AAAWWTWTW`, I would expect the salient test case to be the slowest `T`, not simply the first `W`.

Reply:
This is a tricky one, but since the spec is not quite clear that we break on the first failing case, consistently showing that one would be the simple & straightforward way.
Alternatively we could show the slowest failed case. But this would be a bit inconsistent since it's only due to parallelism that we run any cases other than the first failing one at all :thinking:

Reply:
maybe we don't want `salient` at this point but, multiple methods for, slowest, most memory, first wa, first rte, first tle and then in the UI decide what we want to display?

Reply:
I think more specific error reporting is nice but it can be done in a separate PR.